File: frontend/src/utils/quantityValidation.ts
// Whole-number quantity validation shared by the Product, Purchase Order, GRN and
// Sales forms. Physical stock is counted in whole units, so fractional quantities
// are rejected. Mirrors backend app/utils/quantity_validation.py.

export const WHOLE_QTY_MESSAGE = 'Quantity must be a whole number (decimals are not allowed)';

/** True when value is empty/whole; false when it carries a fractional part. */
export const isWholeQuantity = (value: number | string | null | undefined): boolean => {
  if (value === null || value === undefined || value === '') return true;
  const numeric = Number(value);
  if (!Number.isFinite(numeric)) return false;
  return Number.isInteger(numeric);
};

/**
 * Parse a raw input value to a whole number for controlled quantity inputs.
 * Non-numeric -> 0. Fractional input is truncated so the field can never hold a
 * decimal; submit-time validation still guards against any stray value.
 */
export const parseWholeQuantity = (raw: string): number => {
  const numeric = parseInt(raw, 10);
  return Number.isFinite(numeric) ? numeric : 0;
};

File: backend/app/utils/quantity_validation.py
"""Shared quantity validation.

Physical stock is counted in whole units across the ERP (Product, Purchase Order,
GRN, Sales and their returns). Fractional quantities are rejected at the schema
layer so the same rule is enforced for every write path (create, edit, return).

The validator only *rejects* fractional input; a whole value (e.g. ``5`` or
``5.0``) passes through unchanged, so existing stock/GST/amount calculations are
unaffected.
"""
from __future__ import annotations


def validate_whole_quantity(value, field_label: str = "Quantity"):
    """Return ``value`` unchanged if it is a whole number, else raise ValueError.

    ``None`` is passed through so optional fields keep their optionality; range
    checks (negative/zero) are intentionally left to existing per-schema rules so
    behaviour beyond the decimal check is not altered.
    """
    if value is None:
        return value
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        raise ValueError(f"{field_label} must be a whole number (decimals are not allowed)")
    if numeric != int(numeric):
        raise ValueError(f"{field_label} must be a whole number (decimals are not allowed)")
    return value
